content/blog/write-git-commit-messages-first.md

```diff
@@ -0,0 +1,66 @@
+---
+title: "How — and Why — to Write Commit Messages First"
+featured: false
+description: >
+  Writing your commit messages first keeps you and your pair focused between the feature you are currently trying to implement and whatever current change you are making that could be a smaller, more atomic pushable commit.
+date: 2020-08-19
+tags:
+  - Git
+# Author(s)
+team:
+  - Eric Tsiliacos
+---
+
+You’re halfway through delivering your feature and you decide to take a look at your diff. Doing so gives you a sinking feeling in your stomach, because you see a lot more changes than you were expecting, some of which were refactorings, like renames or structural changes you wish were separated into their own commits. Teasing apart these smaller commits can be messy and would take too much effort at this point. But if you’d tried to preemptively break them apart at random, you’d have run the risk of overengineering your work or creating unnecessary changes.
+
+Writing commit messages first can help pairs navigate between feature delivery and what could be smaller, more atomic commits. It’s analogous to working from “stories” off a backlog or trying to get the next failing test to pass. If you or your pair gets lost in the weeds, pointing back to the story provides a direct and egoless way to get back on track. In test-driven development, pairs stay focused by thinking about small, incremental, falsifiable functionality one failing test at a time. Writing commit messages first can help pairs articulate the space between a feature story and individual tests. 
+
+## How to Write Commit Messages First
+
+You can use a [command line tool](https://github.com/ericTsiliacos/stacker) to keep track of your commit messages, or simply use pencil and paper. The first commit message is easy to write and provides a frictionless start. A lot of thought usually goes into the title or description of a story, so it provides the perfect level of abstraction to communicate the overall “why” for the actual changes needed to deliver the feature. Write down the commit message before changing any code, and work from the assumption that all necessary changes will be committed with this message. When you come across a piece of code that, for example, needs structural changes to make your feature possible, stash or delete your current changes. Then write down the commit message that describes the change you wish was already present so that you can [“make the change easy... then make the easy change.”](https://twitter.com/KentBeck/status/250733358307500032) The new commit message is causally related to the original commit message in the sense that by working from the original commit message, we naturally discover a smaller commit.
```

Comment:
```suggestion
You can use a command line tool such as [Stacker](https://github.com/ericTsiliacos/stacker) to keep track of your commit messages, or simply use pencil and paper. The first commit message is easy to write and provides a frictionless start. A lot of thought usually goes into the title or description of a story, so it provides the perfect level of abstraction to communicate the overall “why” for the actual changes needed to deliver the feature. Write down the commit message before changing any code, and work from the assumption that all necessary changes will be committed with this message. When you come across a piece of code that, for example, needs structural changes to make your feature possible, stash or delete your current changes. Then write down the commit message that describes the change you wish was already present so that you can [“make the change easy... then make the easy change.”](https://twitter.com/KentBeck/status/250733358307500032) The new commit message is causally related to the original commit message in the sense that by working from the original commit message, we naturally discover a smaller commit.
```